### PR TITLE
[5.5] Prefer stricter positive string comparisons.

### PR DESCRIPTION
**Pull request by lucasmichot**
Prefer stricter positive string comparisons.

### CODE REVIEW

**Comment by browner12** [comment]
what are the upsides and possible downsides to this?

from a logic perspective, especially with strings, I think it would be very rare that this would make a difference.

did a quick google, and *maybe/possibly* there is a *slight* performance boost?

https://stackoverflow.com/questions/2401478/why-is-faster-than-in-php

**Comment by lucasmichot** [comment]
@browner12 see https://github.com/laravel/internals/issues/698

File: src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php [hunk 1]
```diff
@@ -172,7 +172,7 @@ public function withCount($relations)
 
             unset($alias);
 
-            if (count($segments) == 3 && Str::lower($segments[1]) == 'as') {
+            if (count($segments) == 3 && Str::lower($segments[1]) === 'as') {
```
Comment (laurencei) [inline src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: Is there a reason this count check not also strict?

Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: This PR is only focusing on stricter comparisons for string, not for int and float
Or else this PR would be be probably to big to review

Comment (laurencei) [inline src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: Fair enough 👍

**Comment by browner12** [comment]
with the strings this change is probably pretty safe. here's the one possible gotcha we could run into (although I don't see it happening).

```php
class Test {

    public function __toString(){
        return 'laravel';
    }
}
```

With the loose comparison:

```php
$test = new Test();

dd($test == 'laravel');    //returns true
```

With the strict comparison:

```php
$test = new Test();

dd($test === 'laravel');    //returns false
```

My guess is this probably doesn't exist in the framework at all, but just a *possible* problem here.

**Comment by lucasmichot** [comment]
Indeed this is an issue that can happen with some class, for instance `HtmlString`, as we really compare that this is a string.
But this remains safe

**Comment by taylorotwell** [comment]
Let's only change the ones where it would *actually* make a difference.

**Comment by xcombelle** [comment]
@taylorotwell I don't understand your closing comment. The idea of the original issue (https://github.com/laravel/internals/issues/698 ) is to remove all the loose comparison in the aim of avoiding introduction of loose comparison in the future in an automated way and maybe catch some bug hiding around.  I don't see how letting loose comparison that don't make difference allow accomplish this goal.

**Pull request closed without merging**